File: gamedex-core/src/main/java/com/github/ykrasik/gamedex/core/javafx/layout/ImageViewResizingPane.java
package com.github.ykrasik.gamedex.core.javafx.layout;

import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.geometry.HPos;
import javafx.geometry.Orientation;
import javafx.geometry.VPos;
import javafx.scene.Node;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.layout.Pane;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;

/**
 * @author Yevgeny Krasik
 */
@Slf4j
public class ImageViewResizingPane extends Pane {
    private static final Double MAX_ENLARGE_RATIO = 2.0;    // FIXME: Property

    private final ImageView imageView;

    public ImageViewResizingPane(@NonNull ImageView imageView) {
        this.imageView = imageView;
        super.getChildren().add(imageView);
    }

    @Override
    protected void layoutChildren() {
        layoutInArea(imageView, 0, 0, imageView.getFitWidth(), imageView.getFitHeight(), 0, HPos.CENTER, VPos.CENTER);
    }

    @Override
    public void resize(double width, double height) {
        final Image image = getImage();
        final double imageWidth = image.getWidth();
        final double imageHeight = image.getHeight();
        final double normalizedWidth = Math.min(width, imageWidth * MAX_ENLARGE_RATIO);
        final double normalizedHeight = Math.min(height, imageHeight * MAX_ENLARGE_RATIO);

        imageView.setFitHeight(normalizedHeight);
        imageView.setFitWidth(normalizedWidth);
        if (Math.abs(normalizedWidth - width) <= 10 &&
            Math.abs(normalizedHeight - height) <= 10) {
            // If there are a few tiny pixel margins between the image and the border (no idea why JavaFX does this, but it does),
            // stretch the image to fill those in.
            imageView.setPreserveRatio(false);
        } else {
            imageView.setPreserveRatio(true);
        }

        super.resize(normalizedWidth, normalizedHeight);
    }

    @Override
    protected double computePrefWidth(double availableHeight) {
        if (availableHeight <= 0) {
            return availableHeight;
        }

        final double ratio = getImageRatio();
        return availableHeight / ratio;
    }

    @Override
    protected double computePrefHeight(double availableWidth) {
        if (availableWidth <= 0) {
            return availableWidth;
        }

        final double ratio = getImageRatio();
        return availableWidth * ratio;
    }

    @Override
    protected double computeMinWidth(double availableHeight) {
        final double prefWidth = computePrefWidth(availableHeight);
        return limitByMaxWidth(prefWidth);
    }

    @Override
    protected double computeMinHeight(double availableWidth) {
        final double prefHeight = computePrefHeight(availableWidth);
        return limitByMaxHeight(prefHeight);
    }

    @Override
    public Orientation getContentBias() {
        return Orientation.VERTICAL;
    }

    @Override
    @Deprecated
    public ObservableList<Node> getChildren() {
        return FXCollections.unmodifiableObservableList(super.getChildren());
    }

    private Image getImage() {
        return imageView.getImage();
    }

    private double getImageRatio() {
        final Image image = getImage();
        final double height = image.getHeight();
        final double width = image.getWidth();
        return height / width;
    }

    private double limitByMaxWidth(double width) {
        final double maxWidth = getMaxWidth();
        return limitIfApplicable(width, maxWidth);
    }

    private double limitByMaxHeight(double height) {
        final double maxHeight = getMaxHeight();
        return limitIfApplicable(height, maxHeight);
    }

    private double limitIfApplicable(double value, double max) {
        return max > 0 ? Math.min(value, max) : value;
    }
}

File: gamedex-app/src/main/kotlin/com/gitlab/ykrasik/gamedex/ui/view/widgets/ImageViewResizingPane.kt
package com.gitlab.ykrasik.gamedex.ui.view.widgets

import com.gitlab.ykrasik.gamedex.common.util.unmodifiable
import javafx.collections.ObservableList
import javafx.geometry.HPos
import javafx.geometry.Orientation
import javafx.geometry.VPos
import javafx.scene.Node
import javafx.scene.image.ImageView
import javafx.scene.layout.Pane

/**
 * User: ykrasik
 * Date: 30/03/2017
 * Time: 08:36
 */
class ImageViewResizingPane(private val imageView: ImageView) : Pane() {

    init {
        children += imageView
    }

    override fun layoutChildren() {
        layoutInArea(imageView, 0.0, 0.0, imageView.fitWidth, imageView.fitHeight, 0.0, HPos.CENTER, VPos.CENTER)
    }

    override fun resize(width: Double, height: Double) {
        val image = imageView.image
        val imageWidth = image.width
        val imageHeight = image.height
        val normalizedWidth = Math.min(width, imageWidth * maxEnlargeRatio)
        val normalizedHeight = Math.min(height, imageHeight * maxEnlargeRatio)

        imageView.fitHeight = normalizedHeight
        imageView.fitWidth = normalizedWidth
        // If there are a few tiny pixel margins between the image and the border (no idea why JavaFX does this, but it does),
        // stretch the image to fill those in.
        imageView.isPreserveRatio = !(Math.abs(normalizedWidth - width) <= 10 && Math.abs(normalizedHeight - height) <= 10)

        super.resize(normalizedWidth, normalizedHeight)
    }

    override fun computePrefWidth(availableHeight: Double): Double =
        if (availableHeight <= 0) {
            availableHeight
        } else {
            availableHeight / imageRatio
        }

    override fun computePrefHeight(availableWidth: Double): Double =
        if (availableWidth <= 0) {
            availableWidth
        } else {
            availableWidth * imageRatio
        }

    override fun computeMinWidth(availableHeight: Double): Double {
        val prefWidth = computePrefWidth(availableHeight)
        return limitByMaxWidth(prefWidth)
    }

    override fun computeMinHeight(availableWidth: Double): Double {
        val prefHeight = computePrefHeight(availableWidth)
        return limitByMaxHeight(prefHeight)
    }

    override fun getContentBias(): Orientation = Orientation.VERTICAL

    @Deprecated("")
    override fun getChildren(): ObservableList<Node> = super.getChildren().unmodifiable()

    private val imageRatio: Double get() {
        val image = imageView.image
        val height = image.height
        val width = image.width
        return height / width
    }

    private fun limitByMaxWidth(width: Double) = limitIfApplicable(width, maxWidth)
    private fun limitByMaxHeight(height: Double) = limitIfApplicable(height, maxHeight)
    private fun limitIfApplicable(value: Double, max: Double) = if (max > 0) Math.min(value, max) else value

    companion object {
        private val maxEnlargeRatio = 2.0    // FIXME: Property
    }
}
